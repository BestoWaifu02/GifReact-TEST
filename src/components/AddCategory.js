import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    console.log("hanlde submit", inputValue);
    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
/* Cuando se envia un form este hace que la pagina se recargue por completo lo cual es malo porque react no recarga la pagina sino componentes, lo que se hace en este caso es quitar el evento de submit llamando la funcion handleSubmit

 Nombre del componentes, al setCategories lo hace obligatorio :) */
