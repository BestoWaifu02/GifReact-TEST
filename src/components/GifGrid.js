import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from "prop-types";

export const GifGrid = ({ category }) => {
  const { data: images, loading } = useFetchGifs(category);

  return (
    <>
      <h3 className="animate__animated animate__fadeIn"> {category} </h3>

      {loading && <p className="animate__animated animate__flash">Loading</p>}

      <div className="card-grid">
        {images.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};
/*MI ENTENDIMIENTO 
    Primero hacemos la peticion fetch con el async y el await, obnetemos los datos de la API y los filtramos para que solo traiga el id,title,url pero esto se saca de la data para despues ser mapeada y el array que te regresa se pone con la const gif la cual se va al setImages del hook y de ahi a la const de images para despues sacar sus propiedades  
  const [images, setImages] = useState([]);
  Este hook hace que no se vuelva a ejecutar la peticion porque tenemos un boton que agrega al contador un numero y por eso recarga el componente pero si usamos este hook la peticion solo se hara 1 vez
  useEffect(() => {
    importamos el getGifts y como retorna una promesa usamos el then 
    getGifts(category).then(setImages)
  }, [category]); 
  cambio de nombre*/

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
