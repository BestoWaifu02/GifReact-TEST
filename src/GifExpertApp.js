import React, { useState } from "react";
import { AddCategory } from "./components/AddCategory";
import { GifGrid } from "./components/GifGrid";
//Props default vacias para que en las pruebas podamos enviar un array
export const GifExpertApp = ({ defaultprops = [] }) => {
  const [categories, setCategories] = useState(defaultprops);

  return (
    <>
      <h2>App By BestoWaifu02 (Diego Robledo Mendoza)</h2>
      <AddCategory setCategories={setCategories} />
      <hr />

      <ol>
        {categories.map((category) => (
          <GifGrid key={category} category={category} />
        ))}
      </ol>
    </>
  );
};
/*  const categories=['samurai X','One punch','SAO']
  const [categories, setcategories] = useState([
    "samurai X",
    "One punch",
    "SAO",
  ]);
  const handleAdd = () => {
      Para agregar un valor al array no se usa el push porque se esta mutando el objeto pero esa es una mala practica, para eso  se usa el useState
    categories.push('Tokyo Ghoul')
  
    Al usar el setcategories de esta forma se jode el estado del anterior porque crea uno nuevo

    setcategories('Tokyo Ghoul')

    El set categories al ser un callback lo puedes manejar de esta forma
    setcategories(cats=>[...cats,'Tokyo Ghoul'])
}; */
