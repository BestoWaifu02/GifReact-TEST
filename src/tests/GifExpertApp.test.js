import { shallow } from "enzyme";
import React from "react";
import "@testing-library/jest-dom";
import { GifExpertApp } from "../GifExpertApp";
import { GifGrid } from "../components/GifGrid";

describe("Pruebas en el componente GifExpertApp", () => {
  test("Debe hacer match con la snapshot ", () => {
    const wrapper = shallow(<GifExpertApp />);
    expect(wrapper).toMatchSnapshot();
  });
  test("Debe de mostrar una lista de categorias", () => {
    const categories = ["dragon ball", "ghalo"];
    const wrapper = shallow(<GifExpertApp defaultprops={categories} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("GifGrid").length).toBe(categories.length);
  });
});
