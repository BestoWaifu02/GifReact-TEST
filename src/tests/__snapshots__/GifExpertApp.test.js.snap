// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente GifExpertApp Debe de mostrar una lista de categorias 1`] = `
<Fragment>
  <h2>
    App By BestoWaifu02 (Diego Robledo Mendoza)
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol>
    <GifGrid
      category="dragon ball"
      key="dragon ball"
    />
    <GifGrid
      category="ghalo"
      key="ghalo"
    />
  </ol>
</Fragment>
`;

exports[`Pruebas en el componente GifExpertApp Debe hacer match con la snapshot  1`] = `
<Fragment>
  <h2>
    App By BestoWaifu02 (Diego Robledo Mendoza)
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol />
</Fragment>
`;
